# Perform API

In [1]:
# Standard Imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Additional Imports
import os, json, math, time
from yelpapi import YelpAPI
from tqdm.notebook import tqdm_notebook

In [2]:
with open('/Users/mkwak/.secret/tmdb_api.json') as f1: #change the path to match YOUR path!!
    login = json.load(f1)
login.keys()

dict_keys(['client-id', 'api-key'])

In [3]:
import tmdbsimple as tmdb
tmdb.API_KEY =  login['api-key']

In [4]:
## make a movie object using the .Movies function from tmdb
movie = tmdb.Movies(603)
## movie objects have a .info dictionary 
info = movie.info()
info

{'adult': False,
 'backdrop_path': '/l4QHerTSbMI7qgvasqxP36pqjN6.jpg',
 'belongs_to_collection': {'id': 2344,
  'name': 'The Matrix Collection',
  'poster_path': '/bV9qTVHTVf0gkW0j7p7M0ILD4pG.jpg',
  'backdrop_path': '/bRm2DEgUiYciDw3myHuYFInD7la.jpg'},
 'budget': 63000000,
 'genres': [{'id': 28, 'name': 'Action'},
  {'id': 878, 'name': 'Science Fiction'}],
 'homepage': 'http://www.warnerbros.com/matrix',
 'id': 603,
 'imdb_id': 'tt0133093',
 'original_language': 'en',
 'original_title': 'The Matrix',
 'overview': 'Set in the 22nd century, The Matrix tells the story of a computer hacker who joins a group of underground insurgents fighting the vast and powerful computers who now rule the earth.',
 'popularity': 71.244,
 'poster_path': '/f89U3ADr1oiB1s9GkdPOEpXUk5H.jpg',
 'production_companies': [{'id': 79,
   'logo_path': '/tpFpsqbleCzEE2p5EgvUq6ozfCA.png',
   'name': 'Village Roadshow Pictures',
   'origin_country': 'US'},
  {'id': 174,
   'logo_path': '/IuAlhI9eVC9Z8UQWOIDdWRKSEJ.png'

In [5]:
import os, time,json
import tmdbsimple as tmdb 
FOLDER = "Data/"
os.makedirs(FOLDER, exist_ok=True)
os.listdir(FOLDER)

['.ipynb_checkpoints',
 'avengers.csv.gz',
 'final_tmdb_data_2000.csv.gz',
 'final_tmdb_data_2001.csv.gz',
 'notebook.csv.gz',
 'title_akas.csv.gz',
 'title_basics.csv.gz',
 'title_ratings.csv.gz',
 'tmdb_api_results_2000.json',
 'tmdb_api_results_2001.json']

In [6]:
def write_json(new_data, filename): 
    """Appends a list of records (new_data) to a json file (filename). 
    Adapted from: https://www.geeksforgeeks.org/append-to-json-file-using-python/"""  
    
    with open(filename,'r+') as file:
        # First we load existing data into a dict.
        file_data = json.load(file)
        ## Choose extend or append
        if (type(new_data) == list) & (type(file_data) == list):
            file_data.extend(new_data)
        else:
             file_data.append(new_data)
        # Sets file's current position at offset.
        file.seek(0)
        # convert back to json.
        json.dump(file_data, file)


In [7]:
def get_movie_with_rating(movie_id):
    """Adapted from source = https://github.com/celiao/tmdbsimple"""
    #get the movie object for the current it
    movie = tmdb.Movies(movie_id)
    
    #save the .info .releases dictionaries
    info = movie.info()
    releases = movie.releases()
    # loop through countries in releases
    for c in releases['countries']:
        # if the counrty abbre ==US
        if c['iso_3166_1'] == 'US':
            ## save a certification key in info with the certifications
            info['certification'] = c['certification']
    return info

## Save Notebook and Avengers as 2 separate .csv.gz files

In [8]:
## Test that the 2 Movie Ids tt0848228 ("The Avengers") 
## work and tt0332280 ("The Notebook").

test_avengers_notebook = ["tt0848228", "tt0332280"]
results_a_n = []
errors_a_n = []

for movie_id in test_avengers_notebook:
    try:
        movie_info = get_movie_with_rating(movie_id)
        results_a_n.append(movie_info)
    except Exception as e:
        errors_a_n.append([movie_id,e])   

pd.DataFrame(results_a_n)


adult                     backdrop_path  \
0  False  /9BBTo63ANSmhC4e6r62OJFuK2GL.jpg   
1  False  /qom1SZSENdmHFNZBXbtJAU0WTlC.jpg   

                               belongs_to_collection     budget  \
0  {'id': 86311, 'name': 'The Avengers Collection...  220000000   
1                                               None   29000000   

                                              genres  \
0  [{'id': 878, 'name': 'Science Fiction'}, {'id'...   
1  [{'id': 10749, 'name': 'Romance'}, {'id': 18, ...   

                                            homepage     id    imdb_id  \
0         https://www.marvel.com/movies/the-avengers  24428  tt0848228   
1  http://www.newline.com/properties/notebookthe....  11036  tt0332280   

  original_language original_title  ...     revenue  runtime  \
0                en   The Avengers  ...  1518815515      143   
1                en   The Notebook  ...   115603229      123   

                                    spoken_languages    status  \
0  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
1  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                                     tagline         title  video  \
0                    Some assembly required.  The Avengers  False   
1  Behind every great love is a great story.  The Notebook  False   

   vote_average vote_count certification  
0         7.706      27655         PG-13  
1         7.880       9868         PG-13  

[2 rows x 26 columns]

In [9]:
## Test that the 2 Movie Ids tt0848228 ("The Avengers") 
## work and tt0332280 ("The Notebook").

test_avengers_notebook = ["tt0848228"]
results_a = []
errors_a = []

for movie_id in test_avengers_notebook:
    try:
        movie_info = get_movie_with_rating(movie_id)
        results_a.append(movie_info)
    except Exception as e:
        errors_a.append([movie_id,e])   

pd.DataFrame(results_a)


adult                     backdrop_path  \
0  False  /9BBTo63ANSmhC4e6r62OJFuK2GL.jpg   

                               belongs_to_collection     budget  \
0  {'id': 86311, 'name': 'The Avengers Collection...  220000000   

                                              genres  \
0  [{'id': 878, 'name': 'Science Fiction'}, {'id'...   

                                     homepage     id    imdb_id  \
0  https://www.marvel.com/movies/the-avengers  24428  tt0848228   

  original_language original_title  ...     revenue  runtime  \
0                en   The Avengers  ...  1518815515      143   

                                    spoken_languages    status  \
0  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                   tagline         title  video  vote_average vote_count  \
0  Some assembly required.  The Avengers  False         7.706      27655   

  certification  
0         PG-13  

[1 rows x 26 columns]

In [10]:
df_avengers = pd.DataFrame(results_a)

In [11]:
df_avengers.head()

adult                     backdrop_path  \
0  False  /9BBTo63ANSmhC4e6r62OJFuK2GL.jpg   

                               belongs_to_collection     budget  \
0  {'id': 86311, 'name': 'The Avengers Collection...  220000000   

                                              genres  \
0  [{'id': 878, 'name': 'Science Fiction'}, {'id'...   

                                     homepage     id    imdb_id  \
0  https://www.marvel.com/movies/the-avengers  24428  tt0848228   

  original_language original_title  ...     revenue  runtime  \
0                en   The Avengers  ...  1518815515      143   

                                    spoken_languages    status  \
0  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                   tagline         title  video  vote_average vote_count  \
0  Some assembly required.  The Avengers  False         7.706      27655   

  certification  
0         PG-13  

[1 rows x 26 columns]

In [12]:
## Save current dataframe to file.
df_avengers.to_csv("Data/avengers.csv.gz",compression='gzip',index=False)
df_avengers = pd.read_csv("Data/avengers.csv.gz", low_memory = False)
df_avengers.head()

adult                     backdrop_path  \
0  False  /9BBTo63ANSmhC4e6r62OJFuK2GL.jpg   

                               belongs_to_collection     budget  \
0  {'id': 86311, 'name': 'The Avengers Collection...  220000000   

                                              genres  \
0  [{'id': 878, 'name': 'Science Fiction'}, {'id'...   

                                     homepage     id    imdb_id  \
0  https://www.marvel.com/movies/the-avengers  24428  tt0848228   

  original_language original_title  ...     revenue  runtime  \
0                en   The Avengers  ...  1518815515      143   

                                    spoken_languages    status  \
0  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                   tagline         title  video  vote_average vote_count  \
0  Some assembly required.  The Avengers  False         7.706      27655   

  certification  
0         PG-13  

[1 rows x 26 columns]

In [13]:
## Test that the 2 Movie Ids tt0848228 ("The Avengers") 
## work and tt0332280 ("The Notebook").

test_avengers_notebook = ["tt0332280"]
results_n = []
errors_n = []

for movie_id in test_avengers_notebook:
    try:
        movie_info = get_movie_with_rating(movie_id)
        results_n.append(movie_info)
    except Exception as e:
        errors_n.append([movie_id,e])   

pd.DataFrame(results_n)


adult                     backdrop_path belongs_to_collection    budget  \
0  False  /qom1SZSENdmHFNZBXbtJAU0WTlC.jpg                  None  29000000   

                                              genres  \
0  [{'id': 10749, 'name': 'Romance'}, {'id': 18, ...   

                                            homepage     id    imdb_id  \
0  http://www.newline.com/properties/notebookthe....  11036  tt0332280   

  original_language original_title  ...    revenue  runtime  \
0                en   The Notebook  ...  115603229      123   

                                    spoken_languages    status  \
0  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                                     tagline         title  video  \
0  Behind every great love is a great story.  The Notebook  False   

   vote_average vote_count certification  
0          7.88       9868         PG-13  

[1 rows x 26 columns]

In [14]:
df_notebook = pd.DataFrame(results_n)

In [15]:
## Save current dataframe to file.
df_notebook.to_csv("Data/notebook.csv.gz",compression='gzip',index=False)
df_notebook = pd.read_csv("Data/notebook.csv.gz", low_memory = False)
df_notebook.head()

adult                     backdrop_path  belongs_to_collection    budget  \
0  False  /qom1SZSENdmHFNZBXbtJAU0WTlC.jpg                    NaN  29000000   

                                              genres  \
0  [{'id': 10749, 'name': 'Romance'}, {'id': 18, ...   

                                            homepage     id    imdb_id  \
0  http://www.newline.com/properties/notebookthe....  11036  tt0332280   

  original_language original_title  ...    revenue  runtime  \
0                en   The Notebook  ...  115603229      123   

                                    spoken_languages    status  \
0  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                                     tagline         title  video  \
0  Behind every great love is a great story.  The Notebook  False   

   vote_average vote_count certification  
0          7.88       9868         PG-13  

[1 rows x 26 columns]

## Extract and Save results that meet criteria in part 1 (2000 or 2001)

In [16]:
# Open saved file and preview again
basics = pd.read_csv("Data/title_basics.csv.gz", low_memory = False)
basics.head()
basics

tconst titleType  \
0       tt0035423     movie   
1       tt0062336     movie   
2       tt0069049     movie   
3       tt0079644     movie   
4       tt0088751     movie   
...           ...       ...   
143047  tt9916170     movie   
143048  tt9916190     movie   
143049  tt9916270     movie   
143050  tt9916362     movie   
143051  tt9916538     movie   

                                             primaryTitle  \
0                                          Kate & Leopold   
1       The Tango of the Widower and Its Distorting Mi...   
2                              The Other Side of the Wind   
3                                           November 1828   
4                                       The Naked Monster   
...                                                   ...   
143047                                      The Rehearsal   
143048                                          Safeguard   
143049                           Il talento del calabrone   
143050                                              Coven   
143051                                Kuambil Lagi Hatiku   

                                    originalTitle  isAdult  startYear  \
0                                  Kate & Leopold        0     2001.0   
1       El Tango del Viudo y Su Espejo Deformante        0     2020.0   
2                      The Other Side of the Wind        0     2018.0   
3                                   November 1828        0     2001.0   
4                               The Naked Monster        0     2005.0   
...                                           ...      ...        ...   
143047                                   O Ensaio        0     2019.0   
143048                                  Safeguard        0     2020.0   
143049                   Il talento del calabrone        0     2020.0   
143050                                   Akelarre        0     2020.0   
143051                        Kuambil Lagi Hatiku        0     2019.0   

        endYear  runtimeMinutes                     genres  
0           NaN             118     Comedy,Fantasy,Romance  
1           NaN              70                      Drama  
2           NaN             122                      Drama  
3           NaN             140                  Drama,War  
4           NaN             100       Comedy,Horror,Sci-Fi  
...         ...             ...                        ...  
143047      NaN              51                      Drama  
143048      NaN              95  Action,Adventure,Thriller  
143049      NaN              84                   Thriller  
143050      NaN              92              Drama,History  
143051      NaN             123                      Drama  

[143052 rows x 9 columns]

In [94]:
YEARS_TO_GET = [2000,2001]

In [95]:
# start of the OUTER LOOP
for YEAR in tqdm_notebook(YEARS_TO_GET, desc= 'YEARS', position = 0):

    # define the JSON file to store results
    JSON_FILE = f'{FOLDER}tmdb_api_results_{YEAR}.json'

    # check if the file exists
    file_exists = os.path.isfile(JSON_FILE)

    # if the file doesn't exist, create it
    if file_exists == False:
        with open(JSON_FILE, 'w') as f:
            json.dump([{'imdb_id':0}],f)

    # save new year as the current df
    df = basics.loc[basics['startYear']==YEAR].copy()

    # save movie ids to list
    movie_ids = df['tconst'].copy() #to_list()
    movie_ids

    # load existing data fro json into new file
    previous_df = pd.read_json(JSON_FILE)
    previous_df

    # filter out any ids that are already in the JSON_FILE
    movie_ids_to_get = movie_ids[~movie_ids.isin(previous_df['imdb_id'])]


    for movie_id in tqdm_notebook(movie_ids_to_get,
                                 desc=f'Movies from {YEAR}',
                                 position = 1,
                                 leave = True):
        try:
            temp = get_movie_with_rating(movie_id)
            write_json(temp,JSON_FILE)
            time.sleep(0.02)
        except Exception as e: 
            errors.append([movie_id, e])
            

    final_year_df = pd.read_json(JSON_FILE)
    final_year_df.to_csv(f"{FOLDER}final_tmdb_data_{YEAR}.csv.gz", compression="gzip", index=False)

YEARS:   0%|          | 0/2 [00:00<?, ?it/s]

Movies from 2000:   0%|          | 0/2716 [00:00<?, ?it/s]

Movies from 2001:   0%|          | 0/2843 [00:00<?, ?it/s]

In [18]:
# Open saved file and preview again
df_1 = pd.read_csv("Data/final_tmdb_data_2000.csv.gz", low_memory = False)
df_1.head()


imdb_id  adult                     backdrop_path belongs_to_collection  \
0          0    NaN                               NaN                   NaN   
1  tt0113026    0.0  /vMFs7nw6P0bIV1jDsQpxAieAVnH.jpg                   NaN   
2  tt0113092    0.0                               NaN                   NaN   
3  tt0116391    0.0                               NaN                   NaN   
4  tt0116748    0.0  /wr0hTHwkYIRC82MwNbhOvqrw27N.jpg                   NaN   

       budget                                             genres homepage  \
0         NaN                                                NaN      NaN   
1  10000000.0  [{'id': 35, 'name': 'Comedy'}, {'id': 10402, '...      NaN   
2         0.0           [{'id': 878, 'name': 'Science Fiction'}]      NaN   
3         0.0  [{'id': 18, 'name': 'Drama'}, {'id': 28, 'name...      NaN   
4         0.0  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...      NaN   

         id original_language   original_title  ... revenue  runtime  \
0       NaN               NaN              NaN  ...     NaN      NaN   
1   62127.0                en  The Fantasticks  ...     0.0     86.0   
2  110977.0                en    For the Cause  ...     0.0    100.0   
3  442869.0                hi             Gang  ...     0.0    152.0   
4  579396.0                hi         Karobaar  ...     0.0    180.0   

                                    spoken_languages    status  \
0                                                NaN       NaN   
1  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
2  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
3  [{'english_name': 'Hindi', 'iso_639_1': 'hi', ...  Released   
4  [{'english_name': 'Hindi', 'iso_639_1': 'hi', ...  Released   

                                         tagline            title  video  \
0                                            NaN              NaN    NaN   
1  Try to remember the first time magic happened  The Fantasticks    0.0   
2   The ultimate showdown on a forbidden planet.    For the Cause    0.0   
3                                            NaN             Gang    0.0   
4                           The Business of Love         Karobaar    0.0   

   vote_average vote_count certification  
0           NaN        NaN           NaN  
1           5.5       22.0           NaN  
2           5.1        8.0           NaN  
3           4.0        1.0           NaN  
4           5.5        2.0           NaN  

[5 rows x 26 columns]

In [98]:
## Just a test to see the above files grabbed the correct columns 
## 

test = ["tt0113092"]
results = []
errors = []

for movie_id in test:
    try:
        movie_info = get_movie_with_rating(movie_id)
        results.append(movie_info)
    except Exception as e:
        errors.append([movie_id,e])   

pd.DataFrame(results)


adult backdrop_path belongs_to_collection  budget  \
0  False          None                  None       0   

                                     genres homepage      id    imdb_id  \
0  [{'id': 878, 'name': 'Science Fiction'}]           110977  tt0113092   

  original_language original_title  ... revenue  runtime  \
0                en  For the Cause  ...       0      100   

                                    spoken_languages    status  \
0  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                                        tagline          title  video  \
0  The ultimate showdown on a forbidden planet.  For the Cause  False   

   vote_average vote_count certification  
0           5.1          8                

[1 rows x 26 columns]

## Exploratory Code

- Load in your csv.gz's of results for each year extracted.
- Concatenate the data into 1 dataframe for the remainder of the analysis.
- Once you have your data from the API, they would like you to perform some light EDA to show:
- How many movies had at least some valid financial information (values > 0 for budget OR revenue)?
- Please exclude any movies with 0's for budget AND revenue from the remaining visualizations.
- How many movies are there in each of the certification categories (G/PG/PG-13/R)?
- What is the average revenue per certification category?
- What is the average budget per certification categor

In [19]:
# Open saved file and preview again
df_2000 = pd.read_csv("Data/final_tmdb_data_2000.csv.gz", low_memory = False)
df_2000.head()


imdb_id  adult                     backdrop_path belongs_to_collection  \
0          0    NaN                               NaN                   NaN   
1  tt0113026    0.0  /vMFs7nw6P0bIV1jDsQpxAieAVnH.jpg                   NaN   
2  tt0113092    0.0                               NaN                   NaN   
3  tt0116391    0.0                               NaN                   NaN   
4  tt0116748    0.0  /wr0hTHwkYIRC82MwNbhOvqrw27N.jpg                   NaN   

       budget                                             genres homepage  \
0         NaN                                                NaN      NaN   
1  10000000.0  [{'id': 35, 'name': 'Comedy'}, {'id': 10402, '...      NaN   
2         0.0           [{'id': 878, 'name': 'Science Fiction'}]      NaN   
3         0.0  [{'id': 18, 'name': 'Drama'}, {'id': 28, 'name...      NaN   
4         0.0  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...      NaN   

         id original_language   original_title  ... revenue  runtime  \
0       NaN               NaN              NaN  ...     NaN      NaN   
1   62127.0                en  The Fantasticks  ...     0.0     86.0   
2  110977.0                en    For the Cause  ...     0.0    100.0   
3  442869.0                hi             Gang  ...     0.0    152.0   
4  579396.0                hi         Karobaar  ...     0.0    180.0   

                                    spoken_languages    status  \
0                                                NaN       NaN   
1  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
2  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
3  [{'english_name': 'Hindi', 'iso_639_1': 'hi', ...  Released   
4  [{'english_name': 'Hindi', 'iso_639_1': 'hi', ...  Released   

                                         tagline            title  video  \
0                                            NaN              NaN    NaN   
1  Try to remember the first time magic happened  The Fantasticks    0.0   
2   The ultimate showdown on a forbidden planet.    For the Cause    0.0   
3                                            NaN             Gang    0.0   
4                           The Business of Love         Karobaar    0.0   

   vote_average vote_count certification  
0           NaN        NaN           NaN  
1           5.5       22.0           NaN  
2           5.1        8.0           NaN  
3           4.0        1.0           NaN  
4           5.5        2.0           NaN  

[5 rows x 26 columns]

In [20]:
# Open saved file and preview again
df_2001 = pd.read_csv("Data/final_tmdb_data_2001.csv.gz", low_memory = False)
df_2001.head()


imdb_id  adult                     backdrop_path belongs_to_collection  \
0          0    NaN                               NaN                   NaN   
1  tt0035423    0.0  /ab5yL8zgRotrICzGbEl10z24N71.jpg                   NaN   
2  tt0079644    0.0  /79axmuH1UGkB7m72jjB9rPff9om.jpg                   NaN   
3  tt0089067    0.0                               NaN                   NaN   
4  tt0114447    0.0                               NaN                   NaN   

       budget                                             genres homepage  \
0         NaN                                                NaN      NaN   
1  48000000.0  [{'id': 35, 'name': 'Comedy'}, {'id': 14, 'nam...      NaN   
2         0.0                     [{'id': 10752, 'name': 'War'}]      NaN   
3         0.0                     [{'id': 35, 'name': 'Comedy'}]      NaN   
4         0.0  [{'id': 53, 'name': 'Thriller'}, {'id': 28, 'n...      NaN   

         id original_language             original_title  ...     revenue  \
0       NaN               NaN                        NaN  ...         NaN   
1   11232.0                en             Kate & Leopold  ...  76019048.0   
2  285529.0                id              November 1828  ...         0.0   
3  210258.0                es  El día de los albañiles 2  ...         0.0   
4  151007.0                en           The Silent Force  ...         0.0   

   runtime                                   spoken_languages    status  \
0      NaN                                                NaN       NaN   
1    118.0  [{'english_name': 'Italian', 'iso_639_1': 'it'...  Released   
2    140.0  [{'english_name': 'Indonesian', 'iso_639_1': '...  Released   
3     90.0  [{'english_name': 'Spanish', 'iso_639_1': 'es'...  Released   
4     90.0  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                                             tagline  \
0                                                NaN   
1  If they lived in the same century they'd be pe...   
2                                                NaN   
3     The laborers are back full of love and laughs.   
4  They left him for dead... They should have fin...   

                       title  video  vote_average vote_count certification  
0                        NaN    NaN           NaN        NaN           NaN  
1             Kate & Leopold    0.0         6.318     1146.0         PG-13  
2              November 1828    0.0         0.000        0.0           NaN  
3  El día de los albañiles 2    0.0         7.200       71.0           NaN  
4           The Silent Force    0.0         5.000        3.0           NaN  

[5 rows x 26 columns]

In [22]:
df_0001 = pd.concat([df_2000,df_2001])

In [23]:
df_0001.head()

imdb_id  adult                     backdrop_path belongs_to_collection  \
0          0    NaN                               NaN                   NaN   
1  tt0113026    0.0  /vMFs7nw6P0bIV1jDsQpxAieAVnH.jpg                   NaN   
2  tt0113092    0.0                               NaN                   NaN   
3  tt0116391    0.0                               NaN                   NaN   
4  tt0116748    0.0  /wr0hTHwkYIRC82MwNbhOvqrw27N.jpg                   NaN   

       budget                                             genres homepage  \
0         NaN                                                NaN      NaN   
1  10000000.0  [{'id': 35, 'name': 'Comedy'}, {'id': 10402, '...      NaN   
2         0.0           [{'id': 878, 'name': 'Science Fiction'}]      NaN   
3         0.0  [{'id': 18, 'name': 'Drama'}, {'id': 28, 'name...      NaN   
4         0.0  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...      NaN   

         id original_language   original_title  ... revenue  runtime  \
0       NaN               NaN              NaN  ...     NaN      NaN   
1   62127.0                en  The Fantasticks  ...     0.0     86.0   
2  110977.0                en    For the Cause  ...     0.0    100.0   
3  442869.0                hi             Gang  ...     0.0    152.0   
4  579396.0                hi         Karobaar  ...     0.0    180.0   

                                    spoken_languages    status  \
0                                                NaN       NaN   
1  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
2  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
3  [{'english_name': 'Hindi', 'iso_639_1': 'hi', ...  Released   
4  [{'english_name': 'Hindi', 'iso_639_1': 'hi', ...  Released   

                                         tagline            title  video  \
0                                            NaN              NaN    NaN   
1  Try to remember the first time magic happened  The Fantasticks    0.0   
2   The ultimate showdown on a forbidden planet.    For the Cause    0.0   
3                                            NaN             Gang    0.0   
4                           The Business of Love         Karobaar    0.0   

   vote_average vote_count certification  
0           NaN        NaN           NaN  
1           5.5       22.0           NaN  
2           5.1        8.0           NaN  
3           4.0        1.0           NaN  
4           5.5        2.0           NaN  

[5 rows x 26 columns]

In [24]:
df_2000.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2193 entries, 0 to 2192
Data columns (total 26 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   imdb_id                2193 non-null   object 
 1   adult                  2192 non-null   float64
 2   backdrop_path          963 non-null    object 
 3   belongs_to_collection  130 non-null    object 
 4   budget                 2192 non-null   float64
 5   genres                 2192 non-null   object 
 6   homepage               90 non-null     object 
 7   id                     2192 non-null   float64
 8   original_language      2192 non-null   object 
 9   original_title         2192 non-null   object 
 10  overview               2006 non-null   object 
 11  popularity             2192 non-null   float64
 12  poster_path            1848 non-null   object 
 13  production_companies   2192 non-null   object 
 14  production_countries   2192 non-null   object 
 15  rele

In [26]:
df_2001.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2262 entries, 0 to 2261
Data columns (total 26 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   imdb_id                2262 non-null   object 
 1   adult                  2261 non-null   float64
 2   backdrop_path          974 non-null    object 
 3   belongs_to_collection  117 non-null    object 
 4   budget                 2261 non-null   float64
 5   genres                 2261 non-null   object 
 6   homepage               141 non-null    object 
 7   id                     2261 non-null   float64
 8   original_language      2261 non-null   object 
 9   original_title         2261 non-null   object 
 10  overview               2091 non-null   object 
 11  popularity             2261 non-null   float64
 12  poster_path            1916 non-null   object 
 13  production_companies   2261 non-null   object 
 14  production_countries   2261 non-null   object 
 15  rele

In [27]:
df_0001.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4455 entries, 0 to 2261
Data columns (total 26 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   imdb_id                4455 non-null   object 
 1   adult                  4453 non-null   float64
 2   backdrop_path          1937 non-null   object 
 3   belongs_to_collection  247 non-null    object 
 4   budget                 4453 non-null   float64
 5   genres                 4453 non-null   object 
 6   homepage               231 non-null    object 
 7   id                     4453 non-null   float64
 8   original_language      4453 non-null   object 
 9   original_title         4453 non-null   object 
 10  overview               4097 non-null   object 
 11  popularity             4453 non-null   float64
 12  poster_path            3764 non-null   object 
 13  production_companies   4453 non-null   object 
 14  production_countries   4453 non-null   object 
 15  rele

In [28]:
bud_non_0 = df_0001['budget'] > 0

In [29]:
rev_non_0 = df_0001['revenue'] > 0

In [36]:
bud_rev_non_0 = df_0001.loc[bud_non_0 | rev_non_0]

In [37]:
bud_rev_non_0.head()

imdb_id  adult                     backdrop_path  \
1   tt0113026    0.0  /vMFs7nw6P0bIV1jDsQpxAieAVnH.jpg   
5   tt0118694    0.0  /n4GJFGzsc7NinI1VeGDXIcQjtU2.jpg   
7   tt0119273    0.0  /f5C03doOWiauu37bToKXtpgP5bS.jpg   
13  tt0120263    0.0  /ynXVuylP8upazjz8lrqb1PEMkdR.jpg   
14  tt0120467    0.0  /knok3mNReKqPTplnnqz7E4dd7mD.jpg   

                                belongs_to_collection      budget  \
1                                                 NaN  10000000.0   
5                                                 NaN    150000.0   
7   {'id': 141086, 'name': 'Heavy Metal Collection...  15000000.0   
13                                                NaN         0.0   
14                                                NaN    120000.0   

                                               genres  \
1   [{'id': 35, 'name': 'Comedy'}, {'id': 10402, '...   
5   [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   
7   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
13  [{'id': 18, 'name': 'Drama'}, {'id': 35, 'name...   
14  [{'id': 18, 'name': 'Drama'}, {'id': 27, 'name...   

                                homepage       id original_language  \
1                                    NaN  62127.0                en   
5   http://www.wkw-inthemoodforlove.com/    843.0                cn   
7                                    NaN  16225.0                en   
13                                   NaN  34070.0                sv   
14                                   NaN  19085.0                en   

                original_title  ...     revenue  runtime  \
1              The Fantasticks  ...         0.0     86.0   
5                         花樣年華  ...  12854953.0     99.0   
7             Heavy Metal 2000  ...         0.0     88.0   
13  Sånger från andra våningen  ...     80334.0     98.0   
14                      Vulgar  ...     14904.0     87.0   

                                     spoken_languages    status  \
1   [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
5   [{'english_name': 'Cantonese', 'iso_639_1': 'c...  Released   
7   [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
13  [{'english_name': 'Russian', 'iso_639_1': 'ru'...  Released   
14  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                                              tagline  \
1       Try to remember the first time magic happened   
5   Feel the heat, keep the feeling burning, let t...   
7   not to survive the fight in the external world...   
13                                                NaN   
14   Everyone loves a clown... some more than others.   

                          title  video  vote_average vote_count certification  
1               The Fantasticks    0.0         5.500       22.0           NaN  
5          In the Mood for Love    0.0         8.112     2008.0            PG  
7              Heavy Metal 2000    0.0         6.134      160.0             R  
13  Songs from the Second Floor    0.0         7.245      220.0           NaN  
14                       Vulgar    0.0         5.400       42.0             R  

[5 rows x 26 columns]

In [38]:
bud_rev_non_0.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 721 entries, 1 to 2239
Data columns (total 26 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   imdb_id                721 non-null    object 
 1   adult                  721 non-null    float64
 2   backdrop_path          573 non-null    object 
 3   belongs_to_collection  109 non-null    object 
 4   budget                 721 non-null    float64
 5   genres                 721 non-null    object 
 6   homepage               85 non-null     object 
 7   id                     721 non-null    float64
 8   original_language      721 non-null    object 
 9   original_title         721 non-null    object 
 10  overview               703 non-null    object 
 11  popularity             721 non-null    float64
 12  poster_path            688 non-null    object 
 13  production_companies   721 non-null    object 
 14  production_countries   721 non-null    object 
 15  relea

- There are 721 entries that have at least a budget or a revenue of over $0

In [55]:
df_0001.groupby('certification')['budget'].describe()

count          mean           std  min  25%         50%  \
certification                                                            
-                1.0  0.000000e+00           NaN  0.0  0.0         0.0   
G               26.0  2.205206e+07  3.236924e+07  0.0  0.0   5676838.0   
NC-17            7.0  0.000000e+00  0.000000e+00  0.0  0.0         0.0   
NR              73.0  1.467125e+06  7.251916e+06  0.0  0.0         0.0   
PG              67.0  2.349627e+07  3.765749e+07  0.0  0.0         0.0   
PG-13          182.0  3.106268e+07  3.388180e+07  0.0  0.0  22000000.0   
R              464.0  9.823285e+06  1.969465e+07  0.0  0.0         0.0   
Unrated          1.0  0.000000e+00           NaN  0.0  0.0         0.0   

                      75%          max  
certification                           
-                     0.0          0.0  
G              30000000.0  115000000.0  
NC-17                 0.0          0.0  
NR                    0.0   60000000.0  
PG             35000000.0  127500000.0  
PG-13          49750000.0  140000000.0  
R              10000000.0  115000000.0  
Unrated               0.0          0.0

- How many movies in each certification category

- There are 26 movies with Ratings G
- There are 67 movies with Ratings PG
- There are 182 movies with Ratings PG-13
- There are 464 movies with Ratings R

- What is average budget for each certification

- Budget of G = 22,052,060
- Budget of PG = 23,496,270
- Budget of PG-13 = 31,062,680
- Budget of R = 9,823,285

In [54]:
df_0001.groupby('certification')['revenue'].describe()

count          mean           std  min  25%         50%  \
certification                                                            
-                1.0  0.000000e+00           NaN  0.0  0.0         0.0   
G               26.0  6.661229e+07  1.246493e+08  0.0  0.0   9367428.0   
NC-17            7.0  1.668286e+03  4.413869e+03  0.0  0.0         0.0   
NR              73.0  2.069719e+06  7.975905e+06  0.0  0.0         0.0   
PG              67.0  5.868451e+07  1.475737e+08  0.0  0.0         0.0   
PG-13          182.0  7.144754e+07  1.232743e+08  0.0  0.0  18773589.5   
R              464.0  1.641628e+07  4.808005e+07  0.0  0.0         0.0   
Unrated          1.0  0.000000e+00           NaN  0.0  0.0         0.0   

                       75%          max  
certification                            
-                     0.00          0.0  
G              77804992.75  579707738.0  
NC-17                 0.00      11678.0  
NR                    0.00   47386287.0  
PG             38378116.00  976475550.0  
PG-13          85042058.00  871368364.0  
R               5200630.00  465361176.0  
Unrated               0.00          0.0

- What is average revenue for each certification

- Revenue of G = 66,612,290
- Revenue of PG = 58,684,510
- Revenue of PG-13 = 71,447,540
- Revenue of R = 16,416,280

In [57]:
## Save current dataframe to file.
df_0001.to_csv("Data/tmdb_results_combined.csv.gz",compression='gzip',index=False)
df_0001_test = pd.read_csv("Data/tmdb_results_combined.csv.gz", low_memory = False)
df_0001_test.head()

imdb_id  adult                     backdrop_path belongs_to_collection  \
0          0    NaN                               NaN                   NaN   
1  tt0113026    0.0  /vMFs7nw6P0bIV1jDsQpxAieAVnH.jpg                   NaN   
2  tt0113092    0.0                               NaN                   NaN   
3  tt0116391    0.0                               NaN                   NaN   
4  tt0116748    0.0  /wr0hTHwkYIRC82MwNbhOvqrw27N.jpg                   NaN   

       budget                                             genres homepage  \
0         NaN                                                NaN      NaN   
1  10000000.0  [{'id': 35, 'name': 'Comedy'}, {'id': 10402, '...      NaN   
2         0.0           [{'id': 878, 'name': 'Science Fiction'}]      NaN   
3         0.0  [{'id': 18, 'name': 'Drama'}, {'id': 28, 'name...      NaN   
4         0.0  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...      NaN   

         id original_language   original_title  ... revenue  runtime  \
0       NaN               NaN              NaN  ...     NaN      NaN   
1   62127.0                en  The Fantasticks  ...     0.0     86.0   
2  110977.0                en    For the Cause  ...     0.0    100.0   
3  442869.0                hi             Gang  ...     0.0    152.0   
4  579396.0                hi         Karobaar  ...     0.0    180.0   

                                    spoken_languages    status  \
0                                                NaN       NaN   
1  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
2  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
3  [{'english_name': 'Hindi', 'iso_639_1': 'hi', ...  Released   
4  [{'english_name': 'Hindi', 'iso_639_1': 'hi', ...  Released   

                                         tagline            title  video  \
0                                            NaN              NaN    NaN   
1  Try to remember the first time magic happened  The Fantasticks    0.0   
2   The ultimate showdown on a forbidden planet.    For the Cause    0.0   
3                                            NaN             Gang    0.0   
4                           The Business of Love         Karobaar    0.0   

   vote_average vote_count certification  
0           NaN        NaN           NaN  
1           5.5       22.0           NaN  
2           5.1        8.0           NaN  
3           4.0        1.0           NaN  
4           5.5        2.0           NaN  

[5 rows x 26 columns]